File: lib/ptabler/software/src/ptabler/steps/read_frame.py
from typing import Mapping

import polars as pl
import polars_pf as ppf
from polars._typing import ParallelStrategy
from polars_pf import CreateTableRequest, PTableColumnSpec, PTableColumnSpecAxis

from .base import PStep, StepContext
from ptabler.common import axis_ref

class ReadFrame(PStep, tag="read_frame"):
    """
    PStep to create a lazy table from provided PFrame.

    Corresponds to the ReadFrameStep defined in the TypeScript type definitions.
    """
    name: str
    """Name to assign to the loaded DataFrame in the tablespace"""
    request: CreateTableRequest
    """Request to create the table"""
    translation: Mapping[str, str]
    """Translation of PFrame column ids into Polars column names"""
    parallel: ParallelStrategy = "auto"
    """Parallel strategy to use for the read"""
    low_memory: bool = False
    """Whether to use low memory mode for the read"""

    def column_ref(self, spec: PTableColumnSpec) -> str:
        if isinstance(spec, PTableColumnSpecAxis):
            return axis_ref(spec.spec)
        if spec.id in self.translation:
            return self.translation[spec.id]
        return spec.id # sliced columns do not require renaming

    def execute(self, ctx: StepContext) -> None:
        if ctx.settings.frame_folder is None:
            raise ValueError("Frame folder is not set")
        
        result: tuple[pl.LazyFrame, ppf.PFrameCache] = ppf.pframe_source(
            ctx.settings.frame_folder,
            self.request,
            spill_path=ctx.settings.spill_folder,
            # column_ref is a function which takes a PTableColumnSpec and returns a polars column name
            # effectively pframe_source applies select with aliases to names returned by column_ref
            column_ref=self.column_ref,
            logger=ppf.logger,
            parallel=self.parallel,
            low_memory=self.low_memory,
        )
        lf, cache = result
        
        ctx.put_table(self.name, lf)
        ctx.chain_task(lambda: cache.dispose())

File: lib/ptabler/schema/src/read_frame.ts
import type { PTableDef, PObjectId } from "@milaboratories/pl-model-common";

/**
 * Represents the configuration for a step that reads data from a PFrame directory into the tablespace.
 * Creates a lazy table from the provided PFrame using polars-pf.
 */
export interface ReadFrameStep {
  /** The type of the step, which is always 'read_frame' for this operation. */
  type: "read_frame";
  /** The name assigned to the loaded DataFrame in the tablespace. */
  name: string;
  /** Request to create the table from the PFrame. */
  request: PTableDef<PObjectId>;
  /**
   * Translation from PFrame column ids (file names) into Polars column names
   * (which will be referenced via pt.col(...)).
   */
  translation: Record<string, string>;
  /**
   * Polars parallel strategy to use for the read. Defaults to 'auto'.
   * @see Parallel option in [scan_parquet](https://docs.pola.rs/api/python/stable/reference/api/polars.scan_parquet.html#polars.scan_parquet) documentation.
   */
  parallel?: "auto" | "columns" | "row_groups" | "prefiltered" | "none";
  /** Whether to use low memory mode for the polars read. Defaults to false. */
  low_memory?: boolean;
}
